### PR TITLE
Three new utils for decluttering CLJS code

### PR DESCRIPTION
**Pull request by MattPutnam**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Product Owner** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/cljs/main/broadfcui/utils.cljs [hunk 1]
```diff
@@ -317,3 +317,23 @@
...
+            (defined-did-mount data)))
+        will-unmount
+        (fn [{:keys [locals] :as data}]
+          (doseq [[event _] listeners-map]
+            (.removeEventListener js/window event (@locals (str "WINDOWLISTENER " event))))
+          (when-let [defined-will-unmount (:component-will-unmount defined-methods)]
+            (defined-will-unmount data)))]
+    (assoc defined-methods
+      :component-did-mount did-mount
+      :component-will-unmount will-unmount)))
```
Comment (davidangb) [inline src/cljs/main/broadfcui/utils.cljs:339]: bonus: get the name (or ref) of the component into here somehow. That would make future debugging nice and easy: you could add log statements inside here that would log e.g. `Table mouseup`.

File: src/cljs/main/broadfcui/persistence.cljs [hunk 2]
```diff
@@ -19,10 +19,28 @@
...
+             :initial (fn [] (merge initial (when version {:v version})))
+             :validator (when version (comp (partial = version) :v))})
+           (when-let [defined (:get-initial-state defined-methods)]
+             (defined data))))
+        component-did-update
+        (fn [{:keys [state] :as data}]
+          (save {:key key :state state :except except :only only})
+          (when-let [defined (:component-did-update defined-methods)]
+            (defined data)))]
+    (assoc defined-methods :get-initial-state get-initial-state :component-did-update component-did-update)))
```
Comment (davidangb) [inline src/cljs/main/broadfcui/persistence.cljs]: is it worthwhile/possible for callers to specify their own validator? I'm wondering if there is a use case for certain components to validate e.g. "as long as localstate's version is greater than or equal to what the component specifies", or "as long as localstate's version is between 3 and 6"

**Review by davidangb — APPROVED** [review]
two optional comments inline, looks good otherwise. Did not test.

File: src/cljs/main/broadfcui/persistence.cljs (hunk 2 above)
Comment (MattPutnam) [reply src/cljs/main/broadfcui/persistence.cljs]: Good idea, done.

File: src/cljs/main/broadfcui/utils.cljs (hunk 1 above)
Comment (MattPutnam) [reply src/cljs/main/broadfcui/utils.cljs:339]: I think you're just going to have to put those logs in your listener functions.

**Review by davidangb — APPROVED** [review]
👍 !

**Pull request merged**